File: src/commands/weekly.ts
import type { HttpClient } from "../api/http.ts";
import { listTimeEntries, type TimeEntry } from "../api/time-entries.ts";
import { listClients } from "../api/clients.ts";
import type { WeeklySummary, ClientSummary } from "../format.ts";

export function getWeekRange(referenceDate: Date): {
  weekStart: string;
  weekEnd: string;
} {
  const d = new Date(referenceDate);
  // Get Monday: getDay() returns 0=Sun, 1=Mon...
  const day = d.getDay();
  const diffToMonday = day === 0 ? -6 : 1 - day;
  const monday = new Date(d);
  monday.setDate(d.getDate() + diffToMonday);

  const friday = new Date(monday);
  friday.setDate(monday.getDate() + 4);

  const fmt = (dt: Date) =>
    dt.getFullYear() +
    "-" +
    String(dt.getMonth() + 1).padStart(2, "0") +
    "-" +
    String(dt.getDate()).padStart(2, "0");

  return { weekStart: fmt(monday), weekEnd: fmt(friday) };
}

export function buildSummary(
  entries: TimeEntry[],
  clientNames: Map<number, string>,
  weekStart: string
): WeeklySummary {
  const monday = new Date(weekStart + "T00:00:00");
  const friday = new Date(monday);
  friday.setDate(monday.getDate() + 4);

  const weekEnd =
    friday.getFullYear() +
    "-" +
    String(friday.getMonth() + 1).padStart(2, "0") +
    "-" +
    String(friday.getDate()).padStart(2, "0");

  // Group by client
  const byClient = new Map<number, number[]>();

  for (const entry of entries) {
    // Use local_started_at (no timezone) to determine day-of-week
    const localDate = entry.local_started_at ?? entry.started_at;
    const entryDate = new Date(localDate);
    // Determine day index (0=Mon, 4=Fri)
    const dayOfWeek = entryDate.getDay();
    const dayIndex = dayOfWeek === 0 ? 6 : dayOfWeek - 1; // 0=Mon...6=Sun

    if (dayIndex < 0 || dayIndex > 4) continue; // Skip weekends

    if (!byClient.has(entry.client_id)) {
      byClient.set(entry.client_id, [0, 0, 0, 0, 0]);
    }
    const daily = byClient.get(entry.client_id)!;
    daily[dayIndex]! += entry.duration;
  }

  // Build client summaries
  const clients: ClientSummary[] = [];
  let grandTotal = 0;

  for (const [clientId, dailySeconds] of byClient) {
    const dailyHours = dailySeconds.map((s) => +(s / 3600).toFixed(2));
    const total = +(dailyHours.reduce((a, b) => a + b, 0)).toFixed(2);
    grandTotal += total;

    clients.push({
      name: clientNames.get(clientId) ?? `Client #${clientId}`,
      daily: dailyHours,
      total,
    });
  }

  // Sort by name
  clients.sort((a, b) => a.name.localeCompare(b.name));
  grandTotal = +(grandTotal.toFixed(2));

  return { weekStart, weekEnd, clients, grandTotal };
}

export async function runWeekly(
  http: HttpClient,
  accountId: string,
  businessId: number,
  options: { weekOf?: string }
): Promise<WeeklySummary> {
  const ref = options.weekOf ? new Date(options.weekOf + "T00:00:00") : new Date();
  const { weekStart, weekEnd } = getWeekRange(ref);

  const [entries, clientNames] = await Promise.all([
    listTimeEntries(http, businessId, {
      startedFrom: weekStart,
      startedTo: weekEnd,
    }),
    listClients(http, accountId),
  ]);

  return buildSummary(entries, clientNames, weekStart);
}

File: internal/commands/weekly.go
package commands

import (
	"fmt"
	"math"
	"sort"
	"strings"
	"time"

	"github.com/spf13/cobra"

	"github.com/hev/freshtime/internal/api"
	"github.com/hev/freshtime/internal/config"
	"github.com/hev/freshtime/internal/format"
)

// WeeklyCmd returns the weekly command.
func WeeklyCmd() *cobra.Command {
	var weekOf string
	var jsonOutput bool

	cmd := &cobra.Command{
		Use:   "weekly",
		Short: "Show weekly time summary grouped by client",
		RunE: func(cmd *cobra.Command, args []string) error {
			return runWeekly(weekOf, jsonOutput)
		},
	}

	cmd.Flags().StringVar(&weekOf, "week-of", "", "Show week containing this date (YYYY-MM-DD)")
	cmd.Flags().BoolVar(&jsonOutput, "json", false, "Output as JSON")

	return cmd
}

func getWeekRange(ref time.Time) (weekStart, weekEnd string) {
	day := ref.Weekday()
	diffToMonday := int(time.Monday - day)
	if day == time.Sunday {
		diffToMonday = -6
	}
	monday := ref.AddDate(0, 0, diffToMonday)
	friday := monday.AddDate(0, 0, 4)

	return monday.Format("2006-01-02"), friday.Format("2006-01-02")
}

func buildSummary(entries []api.TimeEntry, clientNames map[int]string, weekStart string) *format.WeeklySummary {
	monday, _ := time.Parse("2006-01-02", weekStart)
	friday := monday.AddDate(0, 0, 4)
	weekEnd := friday.Format("2006-01-02")

	// Group by client
	byClient := make(map[int][]int) // client_id -> [5]seconds

	for _, entry := range entries {
		localDate := entry.LocalStartedAt
		if localDate == "" {
			localDate = entry.StartedAt
		}
		t, err := time.Parse("2006-01-02T15:04:05", localDate)
		if err != nil {
			// Try ISO with timezone
			t, err = time.Parse(time.RFC3339, localDate)
			if err != nil {
				continue
			}
		}
		dayOfWeek := t.Weekday()
		var dayIndex int
		if dayOfWeek == time.Sunday {
			dayIndex = 6
		} else {
			dayIndex = int(dayOfWeek) - 1 // 0=Mon...6=Sun
		}
		if dayIndex < 0 || dayIndex > 4 {
			continue // skip weekends
		}

		if _, ok := byClient[entry.ClientID]; !ok {
			byClient[entry.ClientID] = make([]int, 5)
		}
		byClient[entry.ClientID][dayIndex] += entry.Duration
	}

	// Build client summaries
	var clients []format.ClientSummary
	var grandTotal float64

	for clientID, dailySeconds := range byClient {
		dailyHours := make([]float64, 5)
		var total float64
		for i, s := range dailySeconds {
			h := math.Round(float64(s)/3600*100) / 100
			dailyHours[i] = h
			total += h
		}
		total = math.Round(total*100) / 100
		grandTotal += total

		name := clientNames[clientID]
		if name == "" {
			name = fmt.Sprintf("Client #%d", clientID)
		}
		clients = append(clients, format.ClientSummary{
			Name:  name,
			Daily: dailyHours,
			Total: total,
		})
	}

	sort.Slice(clients, func(i, j int) bool {
		return strings.ToLower(clients[i].Name) < strings.ToLower(clients[j].Name)
	})
	grandTotal = math.Round(grandTotal*100) / 100

	return &format.WeeklySummary{
		WeekStart:  weekStart,
		WeekEnd:    weekEnd,
		Clients:    clients,
		GrandTotal: grandTotal,
	}
}

func runWeekly(weekOf string, jsonOutput bool) error {
	cfg, err := config.Load()
	if err != nil {
		return err
	}

	http := api.NewClient(cfg)

	ref := time.Now()
	if weekOf != "" {
		ref, err = time.Parse("2006-01-02", weekOf)
		if err != nil {
			return fmt.Errorf("invalid date format: %w", err)
		}
	}
	weekStart, weekEnd := getWeekRange(ref)

	entries, err := api.ListTimeEntries(http, cfg.BusinessID, weekStart, weekEnd)
	if err != nil {
		return err
	}

	clientNames, err := api.ListClients(http, cfg.AccountID)
	if err != nil {
		return err
	}

	summary := buildSummary(entries, clientNames, weekStart)

	if jsonOutput {
		fmt.Println(format.JSON(summary))
	} else {
		fmt.Println(format.Table(summary))
	}
	return nil
}
